File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/response/error/ApplicationLayerError.java
package com.roxiemobile.networkingapi.network.rest.response.error;

import com.annimon.stream.function.Consumer;
import com.roxiemobile.androidcommons.diagnostics.Guard;
import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity;
import com.roxiemobile.networkingapi.network.rest.response.error.nested.ResponseEntityHolder;

import org.jetbrains.annotations.NotNull;

public final class ApplicationLayerError extends AbstractRestApiError {

// MARK: - Construction

    public ApplicationLayerError(@NotNull Throwable cause) {
        super(cause);
        Guard.notNull(cause, "cause is null");
    }

// MARK: - Properties

    /**
     * Returns the type of an error.
     */
    public @NotNull RestApiErrorType getType() {
        return RestApiErrorType.APPLICATION_LAYER;
    }

// MARK: - Methods

    /**
     * Sends a printable representation of this {@code RestApiError}'s description
     * to the consumer.
     */
    @Override
    public void printErrorDescription(@NotNull Consumer<String> consumer) {
        Throwable cause = getCause();

        // Log nested error
        if (cause instanceof ResponseEntityHolder) {
            ResponseEntityHolder entityHolder = (ResponseEntityHolder) getCause();
            ResponseEntity<byte[]> entity = entityHolder.getResponseEntity();

            // Send error description to consumer
            consumer.accept("Request url: " + entity.uri());
            consumer.accept("Status: " + entity.status().value());

            // Good, but we need to go deeper :)
            entityHolder.printErrorDescription(consumer);
        }
        else {
            // Parent processing
            super.printErrorDescription(consumer);
        }
    }
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/response/error/ApplicationLayerError.kt
package com.roxiemobile.networkingapi.network.rest.response.error

import com.annimon.stream.function.Consumer
import com.roxiemobile.networkingapi.network.rest.response.error.nested.ResponseEntityHolder

class ApplicationLayerError(cause: Throwable):
    AbstractRestApiError(cause) {

// MARK: - Methods

    /**
     * Sends a printable representation of this `RestApiError`'s description
     * to the consumer.
     */
    override fun printErrorDescription(consumer: Consumer<String>) {
        val error = this.cause

        // Log nested error
        if (error is ResponseEntityHolder) {

            val responseEntityHolder = error as ResponseEntityHolder
            val responseEntity = responseEntityHolder.responseEntity

            // Send error description to consumer
            consumer.accept("Request link: " + responseEntity.link)
            consumer.accept("Status: " + responseEntity.httpStatus.value())

            // Good, but we need to go deeper :)
            responseEntityHolder.printErrorDescription(consumer)
        }
        else {
            super.printErrorDescription(consumer)
        }
    }
}
